quantum/process_keycode/process_auto_shift.c

```diff
@@ -107,113 +100,83 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
-        if (!autoshift_flags.registered && elapsed < autoshift_timeout) {
-            // Auto-shiftable key is being released before the shift timeout;
-            // simulate the original press then let the usual processing take
-            // care of the release.
-            register_code(keycode);
+        const uint16_t elapsed = TIMER_DIFF_16(now, autoshift_time);
+        if (elapsed < autoshift_timeout) {
+            // Change this and below to 16 if custom keys are added.
+            register_code(autoshift_lastkey);
             autoshift_flags.lastshifted = false;
-            autoshift_flags.registered  = true;
-#    if TAP_CODE_DELAY > 0
-            wait_ms(TAP_CODE_DELAY);
+        } else {
+            // Simulate pressing the shift key.
+            add_weak_mods(MOD_BIT(KC_LSFT));
+            register_code(autoshift_lastkey);
+            autoshift_flags.lastshifted = true;
+#    if !defined(AUTO_SHIFT_NO_REPEAT) && !defined(AUTO_SHIFT_NO_AUTO_REPEAT)
+            if (matrixTrigger) {
+                // Prevents release.
+                return;
+            }
 #    endif
         }
-        autoshift_flush(record->event.time);
-    } else if (record->event.pressed) {
-        if (keycode == KC_LSFT) {
-            // If the user presses shift, auto-shift will not hold shift for
-            // them.
-            autoshift_flags.holding_shift = false;
-        } else if (!autoshift_flags.registered) {
-            // If the key isn't registered yet, it means the timeout hasn't
-            // elapsed, so register the key without additional shifting.
-            autoshift_flush(0);
-            autoshift_lastkey = KC_NO;
-        } else {
-            // The key is registered; flush any shift state for the
-            // non-autoshiftable key.
+
+#    if TAP_CODE_DELAY > 0
+                wait_ms(TAP_CODE_DELAY);
+#    endif
+        unregister_code(autoshift_lastkey);
+        autoshift_flush_shift();
+    }
+    else {
+        // Release after keyrepeat.
+        unregister_code(keycode);
+        if (keycode == autoshift_lastkey) {
+            // This will only fire when the key was the last auto-shiftable
+            // pressed. That prevents aaaaBBBB then releasing a from unshifting
+            // later Bs (if B wasn't auto-shiftable).
             autoshift_flush_shift();
         }
     }
+    // Roll the autoshift_time forward for detecting tap-and-hold.
+    autoshift_time = now;
 }
 
-/** \brief Simulates auto-shifted key presses
+/** \brief Simulates auto-shifted key releases when timeout is hit
  *
  *  Can be called from \c matrix_scan_user so that auto-shifted keys are sent
  *  immediately after the timeout has expired, rather than waiting for the key
  *  to be released.
  */
 void autoshift_matrix_scan(void) {
-    if (!autoshift_flags.in_progress || autoshift_flags.registered) {
-        return;
+    const uint16_t now = timer_read();
```

Comment:
`timer_read()` is an expensive operation (requires synchronization and such depending on the arch), and matrix_scan is called frequently.  Minimize the calls to `timer_read` by making them conditional on autoshift_flags as before.